src/gwdt.c

```diff
@@ -0,0 +1,261 @@
+  neighboring flat pixels by making them point to the same root. The
+  new root of two neighboring pixels is chosen so that it has the
+  greatest DEM difference of the flat pixels in its equivalence
+  class. The first pass determines the correct equivalence classes of
+  each flat pixel. Every chain of labels eventually leads to a root
+  pixel with the greatest DEM difference in its connected
+  components. However, different pixels in the same connected
+  component may still be labeled with different labels. A second pass
+  is required to label each pixel with the root of its equivalence
+  class and the corresponding maximum DEM difference.
+
+  # References
+
+  He, L., Ren, X., Gao, Q., Zhao, X., Yao, B., & Chao, Y. (2017). The
+  connected-component labeling problem: A review of state-of-the-art
+  algorithms. Pattern Recognition, 70, 25-43.
+
+  Hoshen, J., & Kopelman, R. (1976). Percolation and cluster
+  distribution. I. Cluster multiple labeling technique and critical
+  concentration algorithm. Physical Review B, 14(8), 3438.
+
+  Wu, K., Otoo, E., & Suzuki, K. (2009). Optimizing two-pass
+  connected-component labeling algorithms. Pattern Analysis and
+  Applications, 12, 117-135.
+*/
+
+// Insert pixel into the union-find data structure with a given
+// weight. The pixel is the root of a new disjoint set, so it points
+// to itself in the labels array.
+static void new_tree(ptrdiff_t *labels, float *weights, ptrdiff_t pixel,
+                     float weight) {
+  labels[pixel] = pixel;
+  weights[pixel] = weight;
+}
+
+// This struct lets us return a connected component label and the
+// corresponding depth.
+typedef struct {
+  ptrdiff_t label;
+  float weight;
+} labeldepth;
+
+// To find the root of the disjoint set containing `pixel`, follow the
+// pointers in the `labels` array until reaching a root, where the
+// pointer points to itself. Return the label of the root as well as
+// its corresponding weight.
+static labeldepth find_root(ptrdiff_t *labels, float *weights,
+                            ptrdiff_t pixel) {
+  labeldepth r = {pixel, 0.0};
+  while (labels[r.label] != r.label) {
+    r.label = labels[r.label];
+  }
+  r.weight = weights[r.label];
+  return r;
+}
+
+// The trees in the union-find structure can get very tall, requiring
+// many indirections to find the root of a given pixel. One way of
+// managing the tree height is /path compression/, which traverses the
+// tree starting from `pixel` and sets all of the labels and weights
+// and the weights along the way to `root` and `weight`. The next time
+// any of the pixels on the path is accessed, only one indirection is
+// required to find the root.
+static void set_root(ptrdiff_t *labels, float *weights, ptrdiff_t pixel,
+                     ptrdiff_t root, float weight) {
+  while (labels[pixel] != pixel) {
+    ptrdiff_t v = labels[pixel];
+    labels[pixel] = root;
+    weights[pixel] = weight;
+    pixel = v;
+  }
```

Comment:
Why is there a while loop here? Wouldn't `labels[pixel] != pixel` be false after `pixel = v;`?

Reply:
After we run through the loop body once, `labels[pixel] == root` and `pixel` has been replaced with its parent, which was stored in the previous value of `labels[pixel]`. This is a new pixel that we have not yet looked at, so we check again to see if it points at itself. If so, then it is the root of this tree, and we exit the loop. If not, then we set `labels[pixel]` to `root` again and follow the link to its parent and so on. This is the same logic as above in `find_root`, except that we set each label as we visit it.

While trying to figure out how to explain this, I came up with this example, which may help.

Imagine `labels` looks like

```
0 0 3
0 2 6
1 2 5
```

which represents a forest that looks like

```
0 --- 1 --- 2 --- 4
            | 
            | --- 5 --- 8

3 --- 6 --- 7
```

where there are two separate roots (0 and 3). At some point in the algorithm, we call `set_root(labels ,weights, 7, 0, 1.0)` to set the new root of `pixel=7` to `root=0`.

Stepping through the loop:

``` C
labels[7] != 7 // since labels[7] == 6 Enter the loop
v = labels[7]  // v == 6
labels[7] = 0  // Set the label of pixel to root
pixel = v      // pixel == 6, follow the pointer to the parent of pixel

// Second iteration
labels[6] != 6 // since labels[6] == 3
v = labels[6]  // v == 3
labels[6] = 0  // Set label to root
pixel = v      // pixel = 6

// Third iteration
labels[3] == 3 // Stop looping,

// Update the final label
labels[3] = 0  // Line 129
```

Now `labels` looks like

```
0 0 0
0 2 0
1 2 5
```

and the corresponding tree is

```
0 --- 1 --- 2 --- 4
|           | 
|           | --- 5 --- 8
|
| --- 3
| --- 6
| --- 7
```

Reply:
Ah, I see. That makes sense, thanks!